Report argument error when calling variant utility methods with a wrong type

YuriSizov opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/6294.

So, right now we have `floor`, `ceil`, `round`, `abs`, `sign`, `snapped`, and `lerp` which take a variant argument of any kind, but only accept a certain subset of variants. When you pass to them an unsupported type of variant, they err with an "invalid method" error. Which leads to two kinds of messages. If GDScript analyzer can resolve the expression inside of the method call, it reports "Invalid call for function". If it doesn't and we try to run the script, it reports "Invalid call. Nonexistent utility function" (which you have to admit carries a bit of a different meaning, worth rewording perhaps?).

We have an error type for invalid arguments, though, and both `snapped` and `lerp` use it for the second argument, but not for the first one. So that's a simple change. However, this error wants us to supply the expected type, which can only be a single type. `snapped` and `lerp` simply ignore that fact, supply nothing, which means they supply `Variant::NIL`. So the error reads to the users "Cannot convert <type> to Nil", which is very wrong. But then again, there is no support for a purely right way either.

So I try to fix that by passing `Variant::VARIANT_MAX`, which is then converted to string as simply "Variant". Still wrong, but _less_ wrong. Which is the best I can offer without disturbing the core codebase.

New runtime error:
> Invalid type in utility function 'floor'. Cannot convert argument 1 from bool to Variant.

New analyzer error:
> Parse Error: Invalid argument for "floor()" function: argument 1 should be "Variant" but is "bool".

anvilfolk commented:
This one person who doesn't understand much about the core thinks this is a good idea :) at the very least, it does provide slightly more informative error messages. Defs not the method that's the error, it's the arguments :)

vnen commented:
As I mentioned on RC, I think this is fine, but perhaps it is better to use `Variant::NIL` for this case since it's same used in `PropertyInfo` to represent `Variant`. Though this would need a special case when generating the error message.

akien-mga commented:
> As I mentioned on RC, I think this is fine, but perhaps it is better to use `Variant::NIL` for this case since it's same used in `PropertyInfo` to represent `Variant`. Though this would need a special case when generating the error message.

I agree, using a `_MAX` enum value for this still doesn't sit well with me.

Can we confirm that a function will never expected actual `Nil` as its argument type? If so we should indeed be able to consider that `Variant::NIL` for the `expected` type means a generic Variant, as used in PropertyInfo with `PROPERTY_USAGE_NIL_IS_VARIANT`.

If so I guess the message should be changed like this?
```diff
diff --git a/modules/gdscript/gdscript_vm.cpp b/modules/gdscript/gdscript_vm.cpp
index 1ddd54b323..5e6c95a238 100644
--- a/modules/gdscript/gdscript_vm.cpp
+++ b/modules/gdscript/gdscript_vm.cpp
@@ -125,7 +125,8 @@ String GDScriptFunction::_get_call_error(const Callable::CallError &p_err, const
 		} else
 #endif // DEBUG_ENABLED
 		{
-			err_text = "Invalid type in " + p_where + ". Cannot convert argument " + itos(errorarg + 1) + " from " + Variant::get_type_name(argptrs[errorarg]->get_type()) + " to " + Variant::get_type_name(Variant::Type(p_err.expected)) + ".";
+			String expected_type_name = Variant::Type(p_err.expected) == Variant::NIL ? "Variant" : Variant::get_type_name(Variant::Type(p_err.expected));
+			err_text = "Invalid type in " + p_where + ". Cannot convert argument " + itos(errorarg + 1) + " from " + Variant::get_type_name(argptrs[errorarg]->get_type()) + " to " + expected_type_name + ".";
 		}
 	} else if (p_err.error == Callable::CallError::CALL_ERROR_TOO_MANY_ARGUMENTS) {
 		err_text = "Invalid call to " + p_where + ". Expected " + itos(p_err.argument) + " arguments.";
```

YuriSizov commented:
> Can we confirm that a function will never expected actual `Nil` as its argument type?

I can imagine a case where a certain set of arguments could require a particular one to be null. Like, something mutually exclusive. That's what would bother me about hacking it this way: `NIL` is an actual value/kind of Variant, while `MAX` is largely unused in this context, it's a service value. So to me it makes more sense as an "any" value. But I understand how it can sit poorly with someone.

Turning a valid value into a special case would mean that not only do we need to explicitly handle it in GDScript, but also other users of the `Callable` API need to do that, for all method calls, not just variant utilities.

----

A proper way to address this could be to convert `expected` into a bitmask, or add a second field to preserve compatibility. We can use values of `Variant::Type` as powers. That would be a bit more involved, but more correct and flexible for future. Unfortunately, compatibility is a concern because we expose the `CallError` struct one way or another.

YuriSizov committed:
Report argument error when calling variant utility methods with a wrong type

This adds a new call error value when multiple types
of an argument are supported and expected. Support for
this new error type is added to GDScript, with new and
improved error messages.

YuriSizov commented:
Okay, so I've added `Callable::CallError::CALL_ERROR_INVALID_ARGUMENT_MULTI`, which is a new error type for just the occasion. With this code the `expected` field should contain a bitmask of `Variant::Type` values.

This is a somewhat troublesome change, as Variant has many types, and the largest value so far is 37. This means that a 32-bit int cannot fit the bitmask. I had to change the type to `int64_t` to accommodate the mask, but this also comes with a change to GDExtension interface, which breaks compatibility.

..Well, all of it can be an issue really, since support for 64-bit integers can come at an increased cost on some platforms.

But this is a more thorough solution with proper reporting, and I don't see a better way other than a bitmask. Maybe there is a way to compress it, but that would probably also come at a performance cost. So the main question is the compatibility, I guess. Let me know, if this is alright.

----

The new GDScript analyzer error:

> Parse Error: Invalid argument for "floor()" function: argument 1 should be one of "int", "float", "Vector2", "Vector3", "Vector4" but is "String".

The new GDScript runtime error:

> Invalid type in utility function 'floor'. Cannot convert argument 1 from String to one of the following: int, float, Vector2, Vector3, Vector4.

<img width="814" alt="godot windows editor dev x86_64_2023-08-25_20-09-36" src="https://github.com/godotengine/godot/assets/11782833/e1943a26-a77a-48b9-8a02-768f48e77ca1">


----

I've updated all previously touched methods. In a couple of places I made a mistake and the report wasn't in accordance with the checks performed, so I adjusted that. I also adjusted `wrap`, `min`, and `max` where appropriate so the reported error matches the checks performed. Whether the checks themselves are correct, I'm not entirely sure. But that's another issue.

The order of checks is also not great for some cases. For example, executing `snapped` with two different invalid types (say, a string and a bool) will produce a misleading message:

> Parse Error: Invalid argument for "snapped()" function: argument 2 should be "Variant" but is "bool".

I'm not sure why it reports as "Variant", when it should spell out the type of the `x` argument, which is String. But in any case, the check here is about matching types between `x` and `step`, so I had to preserve it as is. Perhaps that can be changed at a later point.

YuriSizov commented:
I've talked with @dsnopek and the change to the GDExtension struct is not possible, as it definitely breaks binary compatibility. There are ways around it, but before going too deep into it, this needs to be approved on the core level to decide if it's even worth the effort.

Maybe there is a way to compactly store a combination of variant types. That would be ideal.

dalexeev commented:
* Partially related to #84368.
